File: sms-frontend/src/services/api/studentService.ts
import apiClient from './client';
import {
  type Student,
  type CreateStudentRequest,
  type UpdateStudentRequest,
  type UpdateStatusRequest,
  type StudentPageResponse,
  type StudentSearchParams,
} from '@/types/student.types';

const STUDENT_BASE_URL = '/students';

export const studentService = {
  /**
   * Create a new student
   */
  async createStudent(data: CreateStudentRequest): Promise<Student> {
    const response = await apiClient.post<Student>(STUDENT_BASE_URL, data);
    return response.data;
  },

  /**
   * Get student by ID
   */
  async getStudentById(id: number): Promise<Student> {
    const response = await apiClient.get<Student>(`${STUDENT_BASE_URL}/${id}`);
    return response.data;
  },

  /**
   * Get student by StudentID
   */
  async getStudentByStudentId(studentId: string): Promise<Student> {
    const response = await apiClient.get<Student>(
      `${STUDENT_BASE_URL}/student-id/${studentId}`
    );
    return response.data;
  },

  /**
   * Search students with pagination and filters
   */
  async searchStudents(params: StudentSearchParams): Promise<StudentPageResponse> {
    const response = await apiClient.get<StudentPageResponse>(STUDENT_BASE_URL, {
      params,
    });
    return response.data;
  },

  /**
   * Update student
   */
  async updateStudent(id: number, data: UpdateStudentRequest): Promise<Student> {
    const response = await apiClient.put<Student>(`${STUDENT_BASE_URL}/${id}`, data);
    return response.data;
  },

  /**
   * Update student status
   */
  async updateStatus(id: number, data: UpdateStatusRequest): Promise<Student> {
    const response = await apiClient.patch<Student>(
      `${STUDENT_BASE_URL}/${id}/status`,
      data
    );
    return response.data;
  },

  /**
   * Delete student (soft delete)
   */
  async deleteStudent(id: number): Promise<void> {
    await apiClient.delete(`${STUDENT_BASE_URL}/${id}`);
  },
};

export default studentService;

File: student-service/src/main/java/com/school/sms/student/application/service/StudentService.java
package com.school.sms.student.application.service;

import com.school.sms.student.application.dto.*;
import org.springframework.data.domain.Pageable;

/**
 * Application service interface for student operations.
 */
public interface StudentService {

    /**
     * Create a new student.
     */
    StudentResponse createStudent(CreateStudentRequest request);

    /**
     * Get student by database ID.
     */
    StudentResponse getStudentById(Long id);

    /**
     * Get student by StudentId (STU-YYYY-XXXXX).
     */
    StudentResponse getStudentByStudentId(String studentId);

    /**
     * Search students with criteria.
     */
    StudentPageResponse searchStudents(StudentSearchCriteria criteria, Pageable pageable);

    /**
     * Update student profile.
     */
    StudentResponse updateStudent(Long id, UpdateStudentRequest request);

    /**
     * Update student status.
     */
    StudentResponse updateStatus(Long id, UpdateStatusRequest request);

    /**
     * Delete student (soft delete by setting status to INACTIVE).
     */
    void deleteStudent(Long id);
}
